libcontainer/init_linux.go

```diff
@@ -492,7 +492,13 @@ func signalAllProcesses(m cgroups.Manager, s os.Signal) error {
 			logrus.Warn(err)
 		}
 	}
-	if err := m.Freeze(configs.Thawed); err != nil {
+
```

Comment:
Wouldn't it be a little weird that the code above for getting pids and actually freezing the container during this function is not recursive but the code for thawing the cgroups is recursive?

Reply:
Oops, I take this back, the GetAllPids function is recursive.  So it's just the Freeze of the current cgroup that is not

Reply:
Yeah, the freeze of the current cgroup isn't recursive, but the unfreeze must be recursive. This is asymmetric, but correct. (Otherwise, the code below might block forever while waiting for the killed processes to become unfrozen, when really we wanted all those PIDs to die now regardless of whether or not they were frozen by a nested cgroup).

Reply:
Or another way of thinking about it: because a task is frozen if any of its parent cgroups are switched to a frozen state,
1. freezing tasks with cgroups is inherently recursive
2. unfreezing with cgroups must be done with manual recursively if you actually wish to unfreeze all the involved tasks (irrespective of the wishes of subcgroups)

And, we do wish to unfreeze tasks against the wishes of any subcgroup, because that's what's required to SIGKILL an entire process tree.

In this particular case it'd be nice if the Linux kernel "respected" SIGKILL on a frozen task (instead of treating it like every other signal, where the task remains alive and blocked until it's unfrozen). But it doesn't, so runC must recursively unfreeze as part of its recursive killing.